### ISSUE TITLE
Storing a daily backup of extension settings on scratch.mit.edu origin

### ISSUE THREAD

**Issue by WorldLanguages** [ScratchAddons/ScratchAddons#6448]
### Why this would be helpful

Apparently it's quite usual for Chrome extensions to get "corrupted" and the user needs to "repair" them. Then, the user may lose all of their settings. We're probably not causing this by ourselves, but we can still offer some type of solution.

Meanwhile, it's quite unusual to get logged out of a website you use frequently, unless on incognito mode or if you manually clear data.

### My suggestion

- If the user visits a Scratch page and it's been more than 24 hours, store a copy of the addon settings on IndexedDB on the scratch.mit.edu origin. (Other extensions may be able to read them)
- Provide a way to rollback to this copy. If storage is apparently empty and this copy exists, notify the user about it so we can be sure they are aware.

### Possible alternatives

Possibly storing a copy in extension IndexedDB (instead of scratch.mit.edu) might be equivalent. I don't have any way to know because it's hard to know why extensions get corrupted. I'm not sure if only `chrome.storage` is affected or if everything stored by the extension is.
I would imagine IndexedDB from all origins are stored in a single file by browsers, so I'd believe it would be equivalent and easier to implement, but I'm not sure.

### Additional context

This happened to griffpatch once (after a bad shutdown, apparently) and we get frequent feedback about it.
Haven't seen evidence of this happening to non-Windows users.

**Comment by WorldLanguages** [comment]
It's quite likely Chromium decides an extension might be "corrupt" if it cannot verify the local version of it exactly matches the Chrome Web Store version. That verification can go wrong for various reasons.

Old screenshot I found the web:
![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/27cbf40e-ee16-4c9f-9a96-7a177be913a6)

**Comment by WorldLanguages** [comment]
Clicking "repair" probably installs the extension "from scratch" again, so that's probably why the storage is reset. It's possible IndexedDB on the extension origin is also cleared through this process.

![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/f51f980e-49d9-4978-a682-00b82f1e5f4e)

**Comment by DNin01** [comment]
That happened to me once, with a different extension. I forgot if I was accidentally messing around with it thinking it was a "from-source" version, though, but it probably wasn't that.

Question is, could Scratch interfere with this storage key or cookie or whatever if it really wanted to or could storing something that's not meant to be used by Scratch, in Scratch, affect Scratch? I'd say this probably isn't a best practice for that reason.

Also, I'm gonna go and back up my settings right now.

**Comment by Secret-chest** [comment]
> Haven't seen evidence of this happening to non-Windows users.

In my 3 years of using this extension on Linux, it hasn't happened once and my PC was shut down by a power surge.

**Comment by GrahamSH-LLK** [comment]
> That happened to me once, with a different extension. I forgot if I was accidentally messing around with it thinking it was a "from-source" version, though, but it probably wasn't that.
> 
> Question is, could Scratch interfere with this storage key or cookie or whatever if it really wanted to or could storing something that's not meant to be used by Scratch, in Scratch, affect Scratch? I'd say this probably isn't a best practice for that reason.
> 
> Also, I'm gonna go and back up my settings right now.

Scratch doesn't touch indexedDB, so everything should be fine.

**Comment by DNin01** [comment]
> > That happened to me once, with a different extension. I forgot if I was accidentally messing around with it thinking it was a "from-source" version, though, but it probably wasn't that.
> > 
> > Question is, could Scratch interfere with this storage key or cookie or whatever if it really wanted to or could storing something that's not meant to be used by Scratch, in Scratch, affect Scratch? I'd say this probably isn't a best practice for that reason.
> > 
> > Also, I'm gonna go and back up my settings right now.
> 
> 
> 
> Scratch doesn't touch indexedDB, so everything should be fine.

Even if it did, I suppose it would be somewhat unlikely for it to be confused by or to go after a database in IndexedDB that it doesn't expect to be there (e.g. wasn't programmed to use), so it will probably be alright.

By that logic, local storage is probably fair game too, right?

**Comment by WorldLanguages** [comment]
We should probably update the privacy policy before working on this

**Comment by DNin01** [comment]
> We should probably update the privacy policy before working on this

Yeah, it does seem a little invasive to store something inside Scratch that's not native to Scratch - enough to deserve a disclaimer. Wouldn't say it's related to privacy, but it seems to me like a good place to say it. Could mention it after we release the update, and it's probably a good idea to notify the user too. Not really sure the best way to write a short description in a way that average users would understand, and in a way that wouldn't make it sound like we are able to do whatever we want. An invasively-toned explanation could result in people uninstalling Scratch Addons.

Could potentially have an opt-out switch if we ever make an "advanced settings" area.

**Comment by DNin01** [comment]
Perhaps there's a way to store it in scratchaddons.com's domain instead, without opening any tabs? Having to grant the host permissions for that could also be suspicious to users as well, though.

When you say, "this happened to griffpatch", did he actually lose his settings?

**Comment by DNin01** [comment]
Could we store a backup in the extension origin's IndexedDB, like we do for Scratch Messaging's cache?

**Comment by WorldLanguages** [comment]
> When you say, "this happened to griffpatch", did he actually lose his settings?

Yes. Off-topic, but his settings weren't that complicated, he only uses a few editor addons.

> Could we store a backup in the extension origin's IndexedDB, like we do for Scratch Messaging's cache?

I mentioned this above, I'm pretty sure that gets cleared which defeats the purpose of the back-up.

**Comment by DNin01** [comment]
By the way, what if another extension or script with access to Scratch was able to modify this database/storage key? Addons usually expect setting values to already be valid/sanitized. If a setting value was modified to an invalid value, could that cause any security issues or undesired behavior? If so, we should remember to validate every setting before loading.
